logstash-core/src/main/java/org/logstash/settings/Setting.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *	http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.logstash.settings;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Predicate;
+
+/**
+ * Root class for all setting definitions.
+ * */
+public class Setting<T> implements Cloneable {
```

Comment:
While it is likely out of scope for this PR, we really have two independent needs for settings: (1) applying user input during process and pipeline start-up and (2) working with the effective values at runtime.

The existing implementation (and therefore the one that you ported) provides objects that address both needs, which means that we have access to mutating explicitly-given settings in places where we have no need to do so.

At some point I would like to be able to have a _minimal_ read-only view, or at least an interface that exposes such, so that we can minimize the risk of explicitly-given settings being overridden by developer error.

Something like:

~~~ java
package co.elastic.logstash.api;

import java.util.function.Supplier;
import java.util.function.UnaryOperator;

public interface Setting<TYPE> {
    /**
     * @return the canonical dot-notation name of this setting
     */
    String getName();

    /**
     * @param valueMapper a mapper that receives the <em>effective</em> value and an
     *                    indication if that is an explicitly-given value, and returns
     *                    a possibly-modified result
     * @return the result of applying the <em>effective</em> value and a flag to the {@code valueMapper}.
     */
    TYPE value(ValueMapper<TYPE> valueMapper);
    
    @FunctionalInterface
    interface ValueMapper<TYPE> {
        TYPE apply(TYPE effectiveValue, boolean isExplicit);
    }
    
    /**
     * @return the explicitly-given value, if set; otherwise the default value
     */
    default TYPE value() {
        return value(((effectiveValue, isExplicit) -> effectiveValue));
    }

    /**
     * @param defaultValueMapper a mapper with which to transform or replace the default value
     * @return the explicitly-given value, if set; otherwise the result of providing
     *         the <em>default</em> value to the {@code defaultValueMapper}.
     */
    default TYPE value(final UnaryOperator<TYPE> defaultValueMapper) {
        return value((effectiveValue, isExplicit) -> isExplicit ? effectiveValue : defaultValueMapper.apply(effectiveValue));
    };

    /**
     * @param defaultValueSupplier a supplier that <em>replaces</em> the default value
     * @return the explicitly-given value, if set; otherwise a default value provided by
     *         the {@code defaultValueSupplier}, <em>ignoring</em> the setting's default.
     */
    default TYPE value(final Supplier<TYPE> defaultValueSupplier) {
        return value((ignoredDefault) -> defaultValueSupplier.get());
    }
}
~~~

We could then implement this interface here in the base-class so that all of our implementations could be _used_ as read-only settings.

Reply:
That would be nice, I think it's would better addressed in a follow up PR. The scope of this PR is to try to do a 1:1 move as mush as possible keeping the existing API. 
Once the full setting hierarchy is moved to Java, with unit tests also, we can do such API refactoring easier.
The base idea, separating the writer from the reader, which is inspired by single-responsibility principle, is a great suggestion 👍 
Let me know if you push to have such change in this PR.